jerry-ext/jerry-context/jerry-context.c

```
@@ -0,0 +1,139 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <string.h>
+#include "jerry-context.h"
+#include "jmem.h"
+
+#define HEAP_BLOCK_SIZE (JERRYX_CONTEXT_SLOTS * sizeof (void *))
+
+typedef struct
+{
+  jerry_user_context_init_cb init_cb;
+  jerry_user_context_deinit_cb deinit_cb;
+} jerryx_context_slot_info;
+
+/* Context-free data */
+
+/* We do not allocate slots after a context has already been created. */
+bool before_first_context = true;
+static int next_index = -1;
+static jerryx_context_slot_info slot_info[JERRYX_CONTEXT_SLOTS] =
+{
+  {
+    NULL, NULL
+  }
+};
+
+/**
+ * Create a new context. This function can be passed to jerry_init_with_user_context () as its second parameter.
+ *
+ * @return the pointer to store as the user context.
+ */
+void *
+jerryx_context_init (void)
+{
+  size_t index = 0;
+  void **return_value = NULL;
+
+  before_first_context = false;
+
+  return_value = (void **) jmem_heap_alloc_block_null_on_error (HEAP_BLOCK_SIZE);
+
+  if (return_value)
+  {
+    memset (return_value, 0, HEAP_BLOCK_SIZE);
+    for (index = 0; index < JERRYX_CONTEXT_SLOTS; index++)
+    {
+      if (slot_info[index].init_cb)
+      {
+        return_value[index] = slot_info[index].init_cb ();
+      }
+    }
```

`return_value == NULL` not being handled / failing silently?
Personally, I'd rather trip an assert in this case.